File: TPs/TP3/solucion.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Forest {
    int t, h, f;
};

vector<Forest> forest_features; // contains (t, h, f) for each forest
vector<vector<vector<int>>> forest; // contains acorns heights for each tree for each forest (forest<tree<acorn>>)

int acorn_pd(int c) {
    int t = forest_features[c].t;
    int h = forest_features[c].h;
    int f = forest_features[c].f;

    vector<vector<int>> M(t, vector<int>(h + 1)); // Memo matrix

    vector<int> M_max(h + 1); // Max at height list, for getting the acorns of the tree that maximices

    for (int tree = 0; tree < t; tree++) {
        for (int acorn_height : forest[c][tree]) {
            M[tree][acorn_height] += 1;
        }
    }

    // Iterate bottom-up
    for (int height = 1; height <= h; height++) {
        for (int tree = 0; tree < t; tree++) {
            if (height - f > 0) { // Verify if i can fly to other tree
                int continue_at_tree = M[tree][height] + M[tree][height - 1];
                int fly_other_tree = M[tree][height] + M_max[height - f];
                M[tree][height] = max(continue_at_tree, fly_other_tree);
            } else { // else, continue
                M[tree][height] += M[tree][height - 1];
            }
            
            // update max acorns at height
            M_max[height] = max(M_max[height], M[tree][height]);
        }
    }
    int max_acorns = *max_element(M_max.begin(), M_max.end());

    return max_acorns;
}

int main() {
    int C;
    cin >> C;

    // Iterate over datasets (forests)
    for (int i = 0; i < C; i++) {

        // First line: t, h, f
        int t, h, f;
        cin >> t >> h >> f;

        forest_features.push_back({t, h, f});

        vector<vector<int>> trees_temp(t); // Resize to hold 't' empty vectors

        // Clear the vector to avoid retaining values from previous dataset
        trees_temp.clear();

        // Iterate over t trees
        for (int j = 0; j < t; j++) {

            // #Acorns on j-th tree
            int a;
            cin >> a;

            // Resize the vector to hold a elements
            trees_temp.push_back(vector<int>(a));

            // Iterate over a acorns
            for (int k = 0; k < a; k++) {
                cin >> trees_temp[j][k];
            }
        }

        forest.push_back(trees_temp);
    }
    int zero; cin >> zero;

    for (int c=0; c<C; c++) {

        int max_acorns = acorn_pd(c);
        cout << max_acorns << endl;
    }

    return 0;
}

File: TPs/TP3/solucion.py
# NOTES
# i'll probably need to save in each tree<acorn> vector the amount of acorns at first position, 
# to know how many iteration i've to do, also to get O(1) accesing that var

class Forest:
    def __init__(self, t, h, f):
        self.t = t
        self.h = h
        self.f = f

def acorns_at_height(t, h, c):
    acorns_at_tree = forest[c][t]
    acorns = 0
    for acorn in acorns_at_tree:
        if acorn == h:
            acorns += 1
    return acorns

def acorn_pd(t, h, f, c, M):
    if h > forest_features[c].h: # base case (got to the top)
        return 0

    if M[t][h] == -1:
        continue_at_tree = acorns_at_height(t, h, c) + acorn_pd(t, h + 1, f, c, M)
        other_trees_max = 0
        for other_tree in range(forest_features[c].t):
            if other_tree != t:
                # Move to other tree, decrease height by 2
                acorn_from_other_tree = acorn_pd(other_tree, h + f, f, c, M)
                other_trees_max = max(other_trees_max, acorn_from_other_tree + acorns_at_height(t, h, c))
        M[t][h] = max(continue_at_tree, other_trees_max)
    return M[t][h]

C = int(input())

forest_features = [] # contains (t, h, f) for each forest
forest = [] # contains acorns heights for each tree for each forest (forest<tree<acorn>>)

# iterate over datasets (forests)
for _ in range(C):
    # first line: t, h, f
    t, h, f = map(int, input().split())
    forest_features.append(Forest(t, h, f))

    trees_temp = []
    # iterate over t trees
    for _ in range(t):
        # #acorns on t-th tree
        acorns_t_tree = list(map(int, input().split()))[1:]  # Skipping the first value which is the number of acorns
        trees_temp.append(acorns_t_tree)

    forest.append(trees_temp)

zero = int(input())

# # Output forest_features
# for feature in forest_features:
#     print(feature.t, feature.h, feature.f)

# # Output forest
# for trees in forest:
#     for tree in trees:
#         print(' '.join(map(str, tree)))

for c in range(C):
    M = [[-1] * (forest_features[c].h + 1) for _ in range(forest_features[c].t)]

    max_acorns = 0

    for tree in range(forest_features[c].t):
        max_acorns_temp = acorn_pd(tree, 1, forest_features[c].f, c, M)
        if max_acorns_temp > max_acorns:
            max_acorns = max_acorns_temp

    print(max_acorns)

